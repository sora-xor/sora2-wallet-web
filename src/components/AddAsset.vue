<template>
  <wallet-base :title="t('addAsset.title')" show-back @back="handleBack">
    <div class="add-asset">
      <s-tabs :value="currentTab" type="rounded" @change="handleChangeTab">
        <s-tab
          v-for="tab in AddAssetTabs"
          :key="tab"
          :label="t(`addAsset.${tab}.title`)"
          :name="tab"
        />
      </s-tabs>
      <component :is="currentTab" @add-asset="handleOpenNotification" />
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Action } from 'vuex-class'

import TranslationMixin from './mixins/TranslationMixin'
import WalletBase from './WalletBase.vue'
import AddAssetSearch from './AddAssetSearch.vue'
import AddAssetCustom from './AddAssetCustom.vue'
import { RouteNames, AddAssetTabs } from '../consts'

@Component({
  components: {
    WalletBase,
    AddAssetSearch,
    AddAssetCustom
  }
})
export default class AddAsset extends Mixins(TranslationMixin) {
  readonly AddAssetTabs = AddAssetTabs

  @Action navigate

  currentTab = AddAssetTabs.Search

  handleChangeTab (value: AddAssetTabs): void {
    this.currentTab = value
  }

  handleBack (): void {
    this.navigate({ name: RouteNames.Wallet })
  }

  handleOpenNotification (): void {
    this.$notify({
      message: this.t('addAsset.success'),
      title: this.t('successText'),
      type: 'success'
    })
  }
}
</script>

<style lang="scss">
.add-asset {
  @include custom-tabs;
}
</style>
