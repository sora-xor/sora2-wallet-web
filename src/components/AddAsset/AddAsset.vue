<template>
  <wallet-base :title="t('addAsset.title')" show-back :reset-focus="currentScreen" @back="handleBack">
    <div class="add-asset">
      <s-tabs v-if="showTabs" :value="currentTab" type="rounded" @change="handleChangeTab">
        <s-tab v-for="tab in AddAssetTabs" :key="tab" :label="t(`addAsset.${tab}.title`)" :name="tab" />
      </s-tabs>
      <keep-alive>
        <component
          :is="currentTab"
          :tokenDetailsPageOpened="tokenDetailsPageOpened"
          @change-visibility="changeVisibility"
        />
      </keep-alive>
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator';

import TranslationMixin from '../mixins/TranslationMixin';
import WalletBase from '../WalletBase.vue';
import AddAssetToken from './AddAssetTokenTab.vue';
import AddAssetNFT from './AddAssetNftTab.vue';
import { RouteNames, AddAssetTabs } from '../../consts';
import { mutation } from '../../store/decorators';
import type { Route } from '../../store/router/types';

@Component({
  components: {
    WalletBase,
    AddAssetToken,
    AddAssetNFT,
  },
})
export default class AddAsset extends Mixins(TranslationMixin) {
  readonly AddAssetTabs = AddAssetTabs;

  @mutation.router.navigate private navigate!: (options: Route) => void;

  currentTab = AddAssetTabs.Token;

  showTabs = true;
  tokenDetailsPageOpened = false;

  get currentScreen(): string {
    return this.currentTab.toString().concat(this.tokenDetailsPageOpened.toString());
  }

  handleChangeTab(value: AddAssetTabs): void {
    this.currentTab = value;
  }

  changeVisibility(): void {
    this.showTabs = false;
    this.tokenDetailsPageOpened = true;
  }

  handleBack(): void {
    if (!this.showTabs) {
      this.showTabs = true;
      this.tokenDetailsPageOpened = false;
      return;
    }

    this.navigate({ name: RouteNames.Wallet });
  }
}
</script>

<style lang="scss">
.add-asset {
  @include custom-tabs;
}
</style>
