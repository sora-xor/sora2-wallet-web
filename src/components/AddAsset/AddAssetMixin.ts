import { Vue, Component, Prop } from 'vue-property-decorator';
import type { AccountAsset, Asset } from '@sora-substrate/util/build/assets/types';

import { state, getter } from '../../store/decorators';
import { AccountAssetsTable } from '@/types/common';

@Component
export default class AddAssetMixin extends Vue {
  @state.account.assets assets!: Array<Asset>;
  @state.account.accountAssets accountAssets!: Array<AccountAsset>;

  @getter.account.accountAssetsAddressTable accountAssetsAddressTable!: AccountAssetsTable;

  @Prop({ default: false, type: Boolean }) tokenDetailsPageOpened!: boolean;

  asset: Nullable<Asset> = null;

  search = '';

  get searchValue(): string {
    return this.search ? this.search.trim().toLowerCase() : '';
  }

  get assetIsAlreadyAdded(): boolean {
    if (!this.searchValue) return false;

    return this.accountAssets.some(
      ({ name = '', symbol = '', address = '' }) =>
        address.toLowerCase() === this.searchValue ||
        symbol.toLowerCase() === this.searchValue ||
        name.toLowerCase() === this.searchValue
    );
  }

  resetSearch(): void {
    this.search = '';
  }

  handleSelectAsset(asset: Asset): void {
    if (asset) {
      this.asset = asset;
      this.$emit('change-visibility');
    }
  }
}
