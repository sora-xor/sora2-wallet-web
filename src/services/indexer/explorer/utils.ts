import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
