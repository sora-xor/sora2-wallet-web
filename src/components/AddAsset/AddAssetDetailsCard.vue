<template>
  <div class="add-asset-details">
    <div class="asset-list-container">
      <div v-for="asset in selectAssets" :key="asset.address">
        <s-card shadow="always" size="small" border-radius="mini" pressed>
          <asset-list-item :asset="asset" :pinnable="false">
            <template #append>
              <s-card size="mini" :status="assetCardStatus(asset)" primary>
                <div class="asset-nature">{{ assetNatureText(asset) }}</div>
              </s-card>
            </template>
          </asset-list-item>
        </s-card>
      </div>
    </div>
    <s-card status="warning" :primary="isCardPrimary" shadow="always" class="add-asset-details_text">
      <div class="p2">{{ t('addAsset.warningTitle') }}</div>
      <div class="warning-text p4">
        {{ warningMessage }}
      </div>
    </s-card>
    <div class="add-asset-details_confirm">
      <s-switch v-model="isConfirmed" :disabled="loading" />
      <span>{{ t('addAsset.understand') }}</span>
    </div>

    <s-button
      class="add-asset-details_action s-typography-button--large"
      type="primary"
      :disabled="!selectAssets.length || !isConfirmed || loading"
      @click="handleAddAssets"
    >
      {{ $tc('addAssetsText', selectAssets.length) }}
    </s-button>
  </div>
</template>

<script lang="ts">
import Theme from '@soramitsu-ui/ui-vue2/lib/types/Theme';
import { Component, Mixins, Prop } from 'vue-property-decorator';

import { api } from '../../api';
import { getter } from '../../store/decorators';
import AssetListItem from '../AssetListItem.vue';
import AddAssetMixin from '../mixins/AddAssetMixin';
import LoadingMixin from '../mixins/LoadingMixin';
import TranslationMixin from '../mixins/TranslationMixin';
import WalletBase from '../WalletBase.vue';

import type { WhitelistIdsBySymbol } from '../../types/common';
import type { Asset, Whitelist } from '@sora-substrate/sdk/build/assets/types';

@Component({
  components: {
    WalletBase,
    AssetListItem,
  },
})
export default class AddAssetDetailsCard extends Mixins(TranslationMixin, LoadingMixin, AddAssetMixin) {
  @getter.account.whitelist whitelist!: Whitelist;
  @getter.account.whitelistIdsBySymbol whitelistIdsBySymbol!: WhitelistIdsBySymbol;

  @Prop({ required: true, type: Array }) readonly selectAssets!: Array<Asset>;
  @Prop({ default: Theme.LIGHT, type: String }) readonly theme!: Theme;
  @Prop({ required: true, type: String }) readonly assetTypeKey!: string;

  isConfirmed = false;

  get isCardPrimary(): boolean {
    return this.theme !== Theme.DARK;
  }

  isWhitelist(asset: Asset): boolean {
    return api.assets.isWhitelist(asset, this.whitelist);
  }

  isBlacklist(asset: Asset): boolean {
    return api.assets.isBlacklist(asset, this.whitelistIdsBySymbol);
  }

  assetCardStatus(asset: Asset): string {
    return this.isWhitelist(asset) ? 'success' : 'error';
  }

  assetNatureText(asset: Asset): string {
    const isWhitelist = this.isWhitelist(asset);
    const isBlacklist = this.isBlacklist(asset);
    if (isWhitelist) {
      return this.t('addAsset.approved');
    }
    if (isBlacklist) {
      return this.t('addAsset.scam');
    }
    return this.t('addAsset.unknown');
  }

  async handleAddAssets(): Promise<void> {
    this.$emit('add');
    this.selectAssets.forEach((asset) => {
      this.addAccountAsset(asset);
    });
  }

  get warningMessage(): string {
    const assetType = this.$tc(`addAsset.assetType.${this.assetTypeKey}`, 1);
    const assetTypePlural = this.$tc(`addAsset.assetType.${this.assetTypeKey}`, this.selectAssets.length);
    const purchaseAssetType =
      this.selectAssets.length === 1
        ? this.$tc('addAsset.warningMessage', 1, { assetType })
        : this.$tc('addAsset.warningMessage', this.selectAssets.length, { assetTypePlural });

    return this.$tc('addAsset.warningMessageText', this.selectAssets.length, {
      assetType,
      assetTypePlural,
      purchaseAssetType,
    });
  }
}
</script>

<style scoped lang="scss">
.add-asset-details {
  & > *:not(:last-child) {
    margin-bottom: #{$basic-spacing-medium};
  }

  .asset-nature {
    font-size: var(--s-font-size-mini);
    font-weight: 300;
    letter-spacing: var(--s-letter-spacing-small);
    line-height: var(--s-line-height-medium);
  }

  &_confirm {
    @include switch-block;

    & {
      padding-top: 0;
      padding-bottom: 0;
    }
  }
  &_action {
    width: 100%;
  }
  &_text {
    color: var(--s-color-status-warning);
  }
  .asset-list-container {
    max-height: 350px;
    overflow-y: auto;
  }
}
</style>
