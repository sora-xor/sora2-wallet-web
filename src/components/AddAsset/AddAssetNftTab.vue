<template>
  <div class="add-asset-nft">
    <div v-if="!tokenDetailsPageOpened" class="add-asset-nft__page">
      <search-input
        autofocus
        v-model="search"
        :placeholder="t(`addAsset.searchInputText`)"
        :maxlength="100"
        @clear="resetSearch"
        class="add-asset-nft__input"
      />
      <asset-list :assets="foundAssets" class="asset-search-list" @click="handleSelectAsset">
        <template #list-empty>
          {{ t(`addAsset.${assetIsAlreadyAdded ? 'alreadyAttached' : 'empty'}`) }}
        </template>
      </asset-list>
    </div>
    <add-asset-details-card v-else :asset="selectedAsset" />
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator';
import type { Asset } from '@sora-substrate/util/build/assets/types';

import AssetList from '../AssetList.vue';
import SearchInput from '../SearchInput.vue';
import AddAssetDetailsCard from './AddAssetDetailsCard.vue';

import AddAssetMixin from '../mixins/AddAssetMixin';
import { api } from '../../api';

@Component({
  components: {
    AssetList,
    SearchInput,
    AddAssetDetailsCard,
  },
})
export default class AddAssetNFT extends Mixins(AddAssetMixin) {
  private get notAddedNftAssets(): Array<Asset> {
    return this.assets.filter((asset) => !(asset.address in this.accountAssetsAddressTable) && api.assets.isNft(asset));
  }

  get foundAssets(): Array<Asset> {
    if (!this.searchValue) return this.notAddedNftAssets;

    return this.getSoughtAssets(this.notAddedNftAssets);
  }

  get assetIsAlreadyAdded(): boolean {
    if (!this.searchValue) return false;

    return this.accountAssets
      .filter((asset) => api.assets.isNft(asset))
      .some(
        ({ name = '', symbol = '', address = '' }) =>
          address.toLowerCase() === this.searchValue ||
          symbol.toLowerCase() === this.searchValue ||
          name.toLowerCase() === this.searchValue
      );
  }
}
</script>

<style scoped lang="scss">
.add-asset-nft {
  &__input {
    margin-top: #{$basic-spacing-medium};
    margin-bottom: #{$basic-spacing-medium};
  }
}
</style>
