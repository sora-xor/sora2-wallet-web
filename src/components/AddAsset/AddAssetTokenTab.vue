<template>
  <div class="add-asset-token">
    <div v-if="!tokenDetailsPageOpened" class="add-asset-token__page">
      <search-input
        autofocus
        v-model="search"
        :placeholder="t(`addAsset.searchInputText`)"
        :maxlength="100"
        @clear="resetSearch"
        class="add-asset-token__search"
      />
      <div class="add-asset-token__switch-btn">
        <s-switch v-model="isVerifiedOnly" :disabled="loading" />
        <span>{{ t(`addAsset.${AddAssetTabs.Token}.switchBtn`) }}</span>
      </div>
      <synthetic-switcher class="add-asset-token__switch-btn" v-model="isSynthsOnly" />
      <asset-list
        :assets="foundAssets"
        class="asset-search-list"
        @click="(asset) => handleSelectAsset(asset)"
        :selectable="isVerifiedOnly"
        :selected="selectedAssets"
      >
        <template #list-empty>
          {{ t(assetIsAlreadyAdded ? 'addAsset.alreadyAttached' : 'addAsset.empty') }}
        </template>
      </asset-list>
      <s-button
        v-if="showAddButton"
        class="add-assets-button"
        type="primary"
        :loading="parentLoading || loading"
        @click="handleAdd"
      >
        {{ t('addAsset.add') }}
      </s-button>
    </div>
    <add-asset-details-card v-else :select-assets="selectedAssets" assetTypeKey="token" />
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator';

import { api } from '../../api';
import { AddAssetTabs, syntheticAssetRegexp } from '../../consts';
import { getter } from '../../store/decorators';
import AssetList from '../AssetList.vue';
import SearchInput from '../Input/SearchInput.vue';
import AddAssetMixin from '../mixins/AddAssetMixin';
import LoadingMixin from '../mixins/LoadingMixin';
import SyntheticSwitcher from '../shared/SyntheticSwitcher.vue';

import AddAssetDetailsCard from './AddAssetDetailsCard.vue';

import type { Asset, Whitelist } from '@sora-substrate/sdk/build/assets/types';

@Component({
  components: {
    AssetList,
    SearchInput,
    AddAssetDetailsCard,
    SyntheticSwitcher,
  },
})
export default class AddAssetToken extends Mixins(LoadingMixin, AddAssetMixin) {
  readonly AddAssetTabs = AddAssetTabs;

  @getter.account.whitelist private whitelist!: Whitelist;
  /** `true` by default cuz we have a lot of assets */
  isVerifiedOnly = true;
  isSynthsOnly = false;

  private get notAddedAssets(): Array<Asset> {
    return this.assets.filter(
      (asset) => !(asset.address in this.accountAssetsAddressTable) && !api.assets.isNft(asset)
    );
  }

  private get prefilteredAssets(): Array<Asset> {
    return this.notAddedAssets.filter((asset) => {
      if (this.isVerifiedOnly && !api.assets.isWhitelist(asset, this.whitelist)) return false;
      if (this.isSynthsOnly && !syntheticAssetRegexp.test(asset.address)) return false;
      return true;
    });
  }

  get foundAssets(): Array<Asset> {
    if (!this.searchValue) return this.prefilteredAssets;

    return this.getSoughtAssets(this.prefilteredAssets);
  }

  get assetIsAlreadyAdded(): boolean {
    if (!this.searchValue) return false;

    return this.accountAssets.some(
      ({ name = '', symbol = '', address = '' }) =>
        address.toLowerCase() === this.searchValue ||
        symbol.toLowerCase() === this.searchValue ||
        name.toLowerCase() === this.searchValue
    );
  }

  get showAddButton(): boolean {
    return this.selectedAssets.length > 0;
  }

  handleAdd() {
    console.info('handle add was called');
    this.$emit('change-visibility');
    // this.selectedAssets.forEach((asset) => this.addAccountAsset(asset));
  }
}
</script>

<style lang="scss">
.asset-search-list {
  @include asset-list($basic-spacing-big);

  .asset {
    padding-left: $basic-spacing-big;
    padding-right: $basic-spacing-big;

    @include focus-outline($withOffset: true);

    &:hover,
    &.selected {
      background-color: var(--s-color-base-background-hover);
      cursor: pointer;
    }

    &-symbol {
      font-size: var(--s-font-size-default);
    }
  }
}
</style>

<style scoped lang="scss">
.add-asset-token {
  margin-top: #{$basic-spacing-medium};

  &__switch-btn {
    display: flex;
    margin-bottom: 10px;
    .s-switch {
      margin-right: 12px;
    }
  }

  &__search {
    margin-bottom: #{$basic-spacing-medium};
  }
}
.add-assets-button {
  width: 100%;
}
</style>
