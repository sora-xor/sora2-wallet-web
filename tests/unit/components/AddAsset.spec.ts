import { useDescribe, useShallowMount } from '../../utils';
import { MOCK_ADD_ASSET } from '../../utils/AddAssetMock';

import AddAsset from '@/components/AddAsset/AddAsset.vue';

useDescribe('AddAsset.vue', AddAsset, () => {
  MOCK_ADD_ASSET.map((item) =>
    it(`[${item.title}]: should be rendered correctly`, () => {
      const wrapper = useShallowMount(AddAsset, {
        data: () => {
          return {
            currentTab: item.currentTab,
          };
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    })
  );
});
