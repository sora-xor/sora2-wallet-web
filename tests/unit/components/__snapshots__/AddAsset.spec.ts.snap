// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAsset.vue [Custom Asset Tab]: should be rendered correctly 1`] = `
<wallet-base-stub
  resetfocus="AddAssetNFTfalse"
  showback="true"
  showheader="true"
  title="Add asset"
  tooltip=""
>
  <div
    class="add-asset"
  >
    <s-tabs-stub
      borderradius="small"
      position="top"
      type="rounded"
      usedesignsystem="true"
      value="AddAssetNFT"
    >
      <s-tab-stub
        label="Tokens"
        name="AddAssetToken"
      />
      <s-tab-stub
        label="NFTs"
        name="AddAssetNFT"
      />
    </s-tabs-stub>
     
    <keep-alive-stub>
      <addassetnft-stub />
    </keep-alive-stub>
  </div>
</wallet-base-stub>
`;

exports[`AddAsset.vue [Search Tab]: should be rendered correctly 1`] = `
<wallet-base-stub
  resetfocus="AddAssetTokenfalse"
  showback="true"
  showheader="true"
  title="Add asset"
  tooltip=""
>
  <div
    class="add-asset"
  >
    <s-tabs-stub
      borderradius="small"
      position="top"
      type="rounded"
      usedesignsystem="true"
      value="AddAssetToken"
    >
      <s-tab-stub
        label="Tokens"
        name="AddAssetToken"
      />
      <s-tab-stub
        label="NFTs"
        name="AddAssetNFT"
      />
    </s-tabs-stub>
     
    <keep-alive-stub>
      <addassettoken-stub />
    </keep-alive-stub>
  </div>
</wallet-base-stub>
`;
